django/researchdata/models.py

```diff
@@ -0,0 +1,397 @@
+    """
+
+
+class SlLetterLocation(SlGeneric):
+    """
+    Select List table: letter person - location
+    Inherits the standard SlGeneric model
+    """
+
+
+class SlLetterType(SlGeneric):
+    """
+    Select List table: letter person - type
+    Inherits the standard SlGeneric model
+    """
+
+
+class SlLetterPersonAppearance(SlGeneric):
+    """
+    Select List table: letter person - appearance
+    Inherits the standard SlGeneric model
+    """
+
+
+class SlLetterPersonRole(SlGeneric):
+    """
+    Select List table: letter person - role
+    Inherits the standard SlGeneric model
+    """
+
+
+class SlLetterEstimatedProportionOfLetter(SlGeneric):
+    """
+    Select List table: letter person - estimated proportion of letter
+    Inherits the standard SlGeneric model
+    """
+
+
+class SlLetterCommentary(SlGeneric):
+    """
+    Select List table: letter person - commentary
+    Inherits the standard SlGeneric model
+    """
+
+
+class SlLetterPersonState(SlGeneric):
+    """
+    Select List table: letter person - state
+    Inherits the standard SlGeneric model
+    """
+
+
+class SlM2MLetterLetterRelationshipType(SlGeneric):
+    """
+    Select List table: type of many to many relationship (letter <-> letter)
+    Inherits the standard SlGeneric model
+    """
+
+
+class SlLetterPersonRelationshipType(SlGeneric):
+    """
+    Select List table: type of many to many relationship (letter <-> person)
+    Inherits the standard SlGeneric model
+    """
+
+
+class SlM2MPersonPersonRelationshipType(SlGeneric):
+    """
+    Select List table: type of many to many relationship (person <-> person)
+    Inherits the standard SlGeneric model
+    """
+
+
+# Main models
+
+
+class Letter(models.Model):
+    """
+    Letter model
+    """
+
+    related_name = "letter"
+
+    title = models.CharField(max_length=255)
+    summary = models.TextField(blank=True, null=True)
+    collection = models.ForeignKey(SlLetterCollection, on_delete=models.SET_NULL, blank=True, null=True)
+    item_number = models.CharField(max_length=255, blank=True, null=True)
+    repository = models.ForeignKey(SlLetterRepository, on_delete=models.SET_NULL, blank=True, null=True)
+    permission_reproduce_text = models.BooleanField(blank=True, null=True)
+    permission_reproduce_image = models.BooleanField(blank=True, null=True)
+    transcription_plain = models.TextField(blank=True, null=True)
+    transcription_normalized = models.TextField(blank=True, null=True)
+    sent_date_year = models.IntegerField(blank=True, null=True,
+                                         validators=[MinValueValidator(1500), MaxValueValidator(2000)])
+    sent_date_month = models.IntegerField(blank=True, null=True,
+                                          validators=[MinValueValidator(1), MaxValueValidator(12)])
+    sent_date_day = models.IntegerField(blank=True, null=True,
+                                        validators=[MinValueValidator(1), MaxValueValidator(31)])
```

Comment:
Is there a reason to not use the Django `DateField` here? It has the added bonus on validating dates - i.e. not allowing 31st February.

Reply:
Yeah this is intentinoal, because they're dealing with historical letters and dates, it's common to not have a complete date. E.g. a letter might say "13th September" but no year, or simply "August" with no day or year, etc.... using a datefield requires a full date to be provided (as I understand) which wouldn't be suitable for this project.